Plane: Differential spoilers support, elevon offset

ethomas997 opened the pull request:
This is a resubmit of PR #2137 , applied to ArduPlane v3.4.1dev as of 2015-10-03.

Fixes ArduPlane differential spoilers when ELEVON_MIXING=0 and ELEVON_OUTPUT>0 (mixing done in flight controller). All flight modes are supported. This functionality is used on flying wings with "split elevons" using two servos on each side. In v3.2-v3.4.0 of ArduPlane, differential spoilers are non-operational (see issue #1032). This patch fixes them so they operate as described in the ArduPlane docs: http://plane.ardupilot.com/wiki/differential-spoilers

Added parameter: DSPOILR_RUD_RATE "Differential spoilers rudder rate"
Sets the amount of deflection that the rudder output will apply to the differential spoilers, as a percentage. The default value of 100 results in full rudder applying full deflection. A value of 0 will result in the differential spoilers exactly following the elevons (no rudder effect).

Added parameter: MIXING_OFFSET
The offset for the Vtail and elevon output mixers, as a percentage. This can be used in combination with MIXING_GAIN to configure how the control surfaces respond to input. The response to aileron or elevator input can be increased by setting this parameter to a positive or negative value. A common usage is to enter a positive value to increase the aileron response of the elevons of a flying wing. The default value of zero will leave the aileron-input response equal to the elevator-input response.

Notes: The new parameter DSPOILR_RUD_RATE allows the rudder effect to be "toned down." Otherwise it can be too easy to apply too much rudder-spoiler control and stall the wing. The new parameter MIXING_OFFSET allows the elevon mixing on a wing to be tuned so that the "aileron" effect is greater than the "elevator" effect. Flying wings are usually pitch-sensitive while being difficult to roll, so this tuning is desirable.

I have successfully tested this patch (on APM2) in a 47" Popwing, in Manual, Stabilized, and GPS-guided flight modes. I've flown it (on APM2) with the patch applied to ArduPlane v3.3.0. I've bench tested the patch, applied to ArduPlane v3.4.0, on Pixhawk.

Source and hex files are also posted here: http://www.etheli.com/APM/ArduPlane_etMod_DiffSpoilersFix

Here is sample set of parameters for testing the new functionality:

DSPOILR_RUD_RATE,75
ELEVON_OUTPUT,2
MIXING_OFFSET,75
RC5_FUNCTION,16
RC5_MAX,1900
RC5_MIN,1100
RC5_REV,1
RC6_FUNCTION,17
RC6_MAX,1900
RC6_MIN,1100
RC6_REV,1

The four servos are plugged into RC1, RC2, RC5 and RC6. One thing to watch out for is that the RC5_MIN, RC5_MAX, RC6_MIN and RC6_MAX parameters can be modified when a radio calibration is performed. If this happens, they should be restored to values like the ones above.

--ET

gmorph commented:
I have built the code and tested and its not working for me.  I will try to determine why.

magicrub commented:
Grant, did you rebase master? Did the original code work?
On Nov 24, 2015 9:28 PM, "Grant Morphett" <notifications@github.com> wrote:

> I have built the code and tested and its not working for me. I will try to
> determine why.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/diydrones/ardupilot/pull/2935#issuecomment-159495441>.
>

gmorph commented:
I had rebased master (see my comment above).
The original code only works with the old ELEVON_MIXING which is why we want this patch which uses the new ELEVON_OUTPUT.
Regardless I now have it working.  I changed a BUNCH of things and it started working.  Reset everything I changed and it still works.  I HATE that.  I have obviously done something with my setup to fix it but I'm not sure what.  At any rate I'm putting it on a physical plane now to check its ok.

gmorph commented:
MIXING_OFFSET is a generic parameter name.  Do we think it should change to something a bit more specific?  I don't have any great ideas - what about - WING_MIXING_OFFSET?  Hmmm, I don't really like that either.  Thoughts?

gmorph commented:
Note the tabbing in the files isn't great.  A lot of the comments are offset from the code.  Would be nice if that could be fixed.

ethomas997 commented:
I called the new parameter MIXING_OFFSET to match up with the existing parameter MIXING_GAIN.

--ET

gmorph commented on ArduPlane/Attitude.cpp:
```diff
@@ -840,13 +844,16 @@ void Plane::set_servos(void)
-					ch2 += abs(channel_rudder->servo_out);
-				    ch4 -= abs(channel_rudder->servo_out);
-				}
+                   //get rudder value and multiply by DSPOILR_RUD_RATE/100:
```
This code block doesn't need to be here as its only called for the old ELEVON_MIXING configuration which we don't really want to change as its legacy code.

gmorph commented on ArduPlane/Attitude.cpp:
```diff
@@ -996,6 +1003,34 @@ void Plane::set_servos(void)
         channel_output_mixer(g.vtail_output, channel_pitch->radio_out, channel_rudder->radio_out);
     } else if (g.elevon_output != MIXING_DISABLED) {
         channel_output_mixer(g.elevon_output, channel_pitch->radio_out, channel_roll->radio_out);
+              //if (both) differential spoilers setup then apply rudder
```
There is a bug in this code.  Because its calling set_radio it doesn't take into account if a channel is reversed i.e. RC5_REV=-1.  
It needs to be changed to call set_servo_out as that then calls calc_pwm which does take into account channel reversal.

gmorph replied:
Whitespace indentation needs fixing on the next 3 lines

gmorph commented:
Hi ET.  First thanks for submitting this PR.  I know its been a long road for you as you have had to rebase a couple of time and resubmit the PR.  We are nearly there!
Are you able to make the code updates I commented on above?  If so I'll test the minute its ready and we will get these code changes into Master as we do need them.
Thanks, Grant.

ethomas997 commented:
Sure, I should be able to work on it over the next few days.

--ET

gmorph commented:
That's great ET - thanks!  I look forward to testing the new code.

ethomas997 commented:
I've updated it to use 'set_servo_out()', which supports the RC REV/MIN/MAX/TRIM settings.  Added constraints to keep the generated values in the 900-2100 range, which matches what is in the existing elevon-mixing code.  Reverted the legacy 'ELEVON_MIXING 1' code back to original (except removed tab chars).

Setting the RC MIN/MAX values for the differential-spoiler channels to 900/2100 should yield matching outputs on the spoilers vs. the elevons.  Here is an updated set of sample parameters for testing:

DSPOILR_RUD_RATE,75
ELEVON_OUTPUT,2
MIXING_OFFSET,75
RC5_FUNCTION,16
RC5_MAX,2100
RC5_MIN,900
RC5_REV,1
RC5_TRIM,1500
RC6_FUNCTION,17
RC6_MAX,2100
RC6_MIN,900
RC6_REV,1
RC6_TRIM,1500

--ET

gmorph commented:
Just a note that we really prefer whitespace changes as a separate commit.  I know it seem's overly fussy but it makes our job much easier when we have to review loads of commits at release time.

ethomas997 commented:
@gmorph Point taken on the whitespace commits.  Do you want me to do an additional commit on this PR that reverses that bit of whitespace mod?  (It just affects about 5 lines of code.)  In general I see this PR as ready.

--ET

magicrub commented:
What's the status of this? @gmorph ? Has it been flown?

kwikius commented:
Looks great, however wouldnt a general purpose HAL mixer be a better approach?

regards
Andy

ethomas997 commented:
Rebased to master 2016-01-23

gmorph commented:
This looks good to go.  I'm out today but will test it this evening and update accordingly.
Thanks, Grant.

gmorph commented:
This ISN'T fixed yet.  I can't see the change to calling set_servo_out().  Do you push your PR correctly?
Thanks, Grant.

ethomas997 commented:
Oops, I think I accidentally swapped in the older version when I rebased.  It should have the updated version now, and I rebased it to the current master (2016-02-11).

--ET

magicrub commented:
It will be great when it goes in. Been a long time! Thanks for taking the time @gmorph to get this through!

gmorph commented on ArduPlane/Attitude.cpp:
```diff
@@ -1041,6 +1045,42 @@ void Plane::set_servos(void)
+                }
+                   //change elevon 1 & 2 positions; constrain min/max:
+                channel_roll->radio_out = constrain_int16(ch1,900,2100);
+                channel_pitch->radio_out = constrain_int16(ch2,900,2100);
```
Rather then the hardcoded values for max and min PWM of 900 and 2100 its better to use the max and min for that channel.  So
channel_roll->radio_min, channel_roll->radio_max
channel_pitch->radio_min, channel_pitch->radio_max

ethomas997 replied:
As I mentioned in my Nov 29th comment, the constraints to keep the generated values in the 900-2100 range match what is in the existing elevon-mixing code (see line 698 with the comment "scale for a 1500 center and 900..2100 range, symmetric").  The channel-min/max values are applied later in the 'set_servo_out()' calls, yes?

gmorph commented on ArduPlane/Attitude.cpp:
```diff
@@ -1041,6 +1045,42 @@ void Plane::set_servos(void)
+                   //constrain min/max for intermediate dspoiler positions:
+                ch3 = constrain_int16(ch3,900,2100);
+                ch4 = constrain_int16(ch4,900,2100);
+            }
```
Same here as above

gmorph commented on ArduPlane/Attitude.cpp:
```diff
@@ -1041,6 +1045,42 @@ void Plane::set_servos(void)
+            int16_t ruddVal = (int16_t)((int32_t)(channel_rudder->servo_out) *
+                                         g.dspoiler_rud_rate / (SERVO_MAX/5));
+            if (ruddVal != 0) {   //if nonzero rudder then apply to spoilers
+                int16_t ch1 = ch3;          //elevon 1
```
whitespace needs fixing in this comment on this line and the next

gmorph commented on ArduPlane/Attitude.cpp:
```diff
@@ -1041,6 +1045,42 @@ void Plane::set_servos(void)
+                    ch2 += ruddVal;
+                    ch4 -= ruddVal;
+                }
+                   //change elevon 1 & 2 positions; constrain min/max:
```
whitespace indentation needs fixing

gmorph commented on ArduPlane/Attitude.cpp:
```diff
@@ -1041,6 +1045,42 @@ void Plane::set_servos(void)
+                   RC_Channel_aux::function_assigned(RC_Channel_aux::k_dspoiler2)) {
+            int16_t ch3 = channel_roll->radio_out;    //diff spoiler 1
+            int16_t ch4 = channel_pitch->radio_out;   //diff spoiler 2
+                   //convert rudder-servo output (-4500 to 4500) to PWM
```
indentation needs fixing

gmorph commented on ArduPlane/Attitude.cpp:
```diff
@@ -1041,6 +1045,42 @@ void Plane::set_servos(void)
+                   //convert rudder-servo output (-4500 to 4500) to PWM
+                   // offset value and multiply by DSPOILR_RUD_RATE/100:
+            int16_t ruddVal = (int16_t)((int32_t)(channel_rudder->servo_out) *
+                                         g.dspoiler_rud_rate / (SERVO_MAX/5));
```
This line is going to divide SERVO_MAX (which is 4500) by 5 which is 900 by  my maths.  That means the comment above is incorrect doesn't it?

ethomas997 replied:
It's a reduction of the formula:  rudder->servo_out * 500 / SERVO_MAX * dspoiler_rud_rate/100.  I've added this to the comment.

ethomas997 commented:
I've processed modified sections through 'astyle' (to conform comment indentations, etc), and rebased to current master 2016-02-20.

--ET

marvinlange commented:
Could this possibly Be used for Differential thrust in twin engine Wing

iskess commented:
See #1861 for twin motor plane

marvinlange commented:
Until this is finish if I just added some flap like vertical spoilers to my the verticals on my flying wings  tips will the apm use them as rudder yaw when elvon mixing is enabled ? or enabled dose it just forget about yaw control ?
![2016-03-08 00 06 22](https://cloud.githubusercontent.com/assets/17716636/13592685/a0a54e5a-e4c2-11e5-89a1-0dbea2efab41.jpg)
![2016-02-16 14 54 38](https://cloud.githubusercontent.com/assets/17716636/13592717/e46b7376-e4c2-11e5-8d15-98246a81f64d.jpg)

ethomas997 commented:
Rebased to current master (2016-04-26) and changes put into single commit.

@gmorph As far as I can tell this is ready for merging.  Please let me know if something else needs to be done with it on my end.

--ET

magicrub commented:
@ethomas997  you've been a champ keeping up on this PR with the rebases. Sorry it's taking so long. @tridge Let's try and get this into v3.6.0 if we can.

magicrub commented:
@ethomas997 are you able to do a rebase soon? Perhaps also a wiki entry? We'd like to get this merged ASAP

gmorph commented:
I have rebased on master, made the necessary fixes and tested it again on my plane on the ground.  It works correctly.  I submitted a new PR for it https://github.com/ArduPilot/ardupilot/pull/4283 but you will still get the commit credit Thomas.  Thanks for your patience.
Thanks, Grant.

magicrub commented:
merged, thanks! (https://github.com/ArduPilot/ardupilot/commit/7d824247af60dd9d3ce97d4d0f2cc19f25a619ef)

WickedShell commented:
There is a report of someone having trouble getting their elevons to work correctly with plane 3.6 but that they work fine with plane 3.5 so I'm wondering if this is related? http://discuss.ardupilot.org/t/3-6-elevons-mixing-gain-mixing-offset-cant-get-proper-throw/9648

magicrub commented:
Possibly. Should be able to see with the other guys params.
On Jun 12, 2016 1:53 PM, "WickedShell" <notifications@github.com> wrote:

> There is a report of someone having trouble getting their elevons to work
> correctly with plane 3.6 but that they work fine with plane 3.5 so I'm
> wondering if this is related?
> http://discuss.ardupilot.org/t/3-6-elevons-mixing-gain-mixing-offset-cant-get-proper-throw/9648
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/ArduPilot/ardupilot/pull/2935#issuecomment-225459709>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AEj7Gz25fsA0TVXspu89l_wMlnj8_Fspks5qLHHlgaJpZM4GIi6l>
> .
>

jgilbertfpv commented:
I got here by following a different post asking about what i was looking for as well, but i don't seem to see any relation. i have differential spoilers and crowing one side at a time with yaw working on a fx-79 buffalo flying wing, now how do i also get both sides to crow for landing flaps?
